specs/SharedBuffer.md

```diff
@@ -0,0 +1,333 @@
+Shared Buffer Between Native Application Code and Script
+===
+
+# Background
+For some advanced scenarios, there is a need to exchange large amounts of data
+between the WebView2 application process and trusted web pages that are considered
+as part of the app. Some examples:
+- Web page generates a large amount of data, and passes it to the native side to be
+  further processed or fed to other parts of the app or OS. For example, the web page
+  generates 100MBs of high DPI images to be printed and needs to pass that to the native
+  code to print. See https://github.com/MicrosoftEdge/WebView2Feedback/issues/89.
+- Native side generates a large amount of data for the web side to consume. The data
+  might or might not come directly from files. For example the native side has generated
+  terrabytes of data to produce different graphs on the web side.
+  See https://github.com/MicrosoftEdge/WebView2Feedback/issues/1005.
+
+To support these scenarios, we are adding an Edge WebView2 API to support sharing
+buffers between the WebView2 host app process and WebView2 renderer process, based
+on shared memory from the OS.
+
+The application code can use the APIs to create a shared buffer object, and share to
+scripts as [ArrayBuffer](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/ArrayBuffer) object.
+Then both the native application code and the script will be able to access the same memory.
+
+# Conceptual pages (How To)
+
+Besides using the memory address directly, the shared buffer object can be accessed
+from native side via an IStream* object that you can get from the shared buffer object.
+
+When the application code calls `PostSharedBufferToScript`, the script side will
+receive a `SharedBufferReceived` event containing the buffer as an `ArrayBuffer` object.
+After receiving the shared buffer object, it can access it the same way as any other
+ArrayBuffer object, including transering to a web worker to process the data on
+the worker thread.
+
+As shared buffer normally represent a large memory, instead of waiting for garbage
+collection to release the memory along with the owning objects, the application
+should try to release the buffer as soon as it doesn't need access to it.
+This can be done by calling Close() method on the shared buffer object, or
+`chrome.webview.releaseBuffer` from script.
+
+As the memory could contain sensitive information and corrupted memory could crash
+the application, the application should only share buffer with trusted sites.
+
+The application can use other messaging channel like `PostWebMessageAsJson` and
+`chrome.webview.postMessage` to inform the other side the desire to have a shared
+buffer and the status of the buffer (like data is produced or consumed).
+
+# Examples
+
+The example below illustrates how to send data from application to script for one time read only consumption.
+
```

Comment:
If I create one CoreWebView2SharedBuffer and pass it to two PostSharedBufferToScript calls, is that an application error?

Reply:
The app should normally not do that. But it is not necessarily a bug. The app could post the buffer for script to consume the data, script releases the buffer after consuming the data. Later on, the app fill buffer with different data and post it again to the script.
Implementation wise, when it is posted twice, we will share the buffer to the script side twice, and creates 2 views of the same memory, represented by 2 ArrayBuffer objects, each can be separately released.
As the buffer can be posted to multiple webviews and iframes for sharing data among multiple script contexts, it is not easy to limit how many times the buffer can be posted.